### ISSUE TITLE
Investigate moving ModelCache into native memory manager

### ISSUE THREAD

**Issue by jmazanec15** [opensearch-project/k-NN#375]
Currently, we have a separate cache for the models and other native memory objects. On a miss, the model cache will pull the model from the model index. During indexing, a model is used to initialize a native library index in native memory. For each new segment, the model is copied to native memory, used to initialize the new index file, and deallocated. This memory pattern is inefficient. The model should just be copied into native memory once.

TLDR: Move model caching to Native Memory Manager and remove ModelCache for more efficient use of resources.

**Comment by jmazanec15** [comment]
Closing in favor of #404
